### PR TITLE
[Merged by Bors] - activation: update NIPostBuilder to use state DB instead of files

### LINKED ISSUE #5206: Add second database for local state
## Description
The node keeps track of the network state inside its database. Additionally it has local state - including but not limited to generated PoST proofs, challenges submitted to PoET, etc. Local state currently is primarily stored as binary data in files on the file system. The goal of this issue is to add a second database for local state and migrate existing data to there.

## Acceptance criteria
- Add second database (`node.sql`) in the same directory as the global state database (`state.sql`)
- Migrate existing state from `activation.Builder` and `activation.NIPostBuilder` to the local database.

### PR DESCRIPTION
**Pull request by fasmat**
## Motivation
Closes #5206

## Changes
- Migrations for `localsql` were extended
  - All migrations are now in code to ensure that data is moved from disk into DB when migrating, failing to do so will result in the DB migration not completing
  - All Go Migrations now have a SQL migration that does the same DB changes, but doesn't migrate data from disk into the DB. This is helpful for when no data needs to be migrated (e.g. in tests). Additionally the Go Migrations can just be deleted in the future when most people are running v1.3.0 or newer versions.
  - Go migrations have tests to ensure data is migrated properly and SQL migrations match the Go Migrations (produce the same Schema after migrating)
- NIPostBuilder now stores its state inside the local database instead of on disk
  - Added new `nipost` localsql functions that allow to store and fetch the data
- Deflake some tests in the `activation/e2e` package by avoiding the spawned PoET from having slightly different timings than the component under test.

## Test Plan
- existing tests have been updated to work with new code
- added tests for new code

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed
- [x] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: activation/activation_test.go
Comment (fasmat) [inline activation/activation_test.go:402]: TODO

File: activation/activation_test.go
Comment (fasmat) [inline activation/activation_test.go:1387]: TODO: add migration code + tests for nipost builder state

File: activation/nipost.go
Comment (fasmat) [inline activation/nipost.go:183]: Opinions? Alternatively I could revert `nb.submitPoetChallenges` to return the list of registrations

File: activation/nipost.go
Comment (fasmat) [inline activation/nipost.go:193]: same as above; before if the state could not be read from disk it would just assume that there were not registrations yet.

File: activation/nipost.go
Comment (fasmat) [inline activation/nipost.go:221]: `nb.persistState()` did also only log on warning that it couldn't persist but would continue.

File: activation/nipost.go
Comment (fasmat) [inline activation/nipost.go:225]: nb.persistState() did also only log on warning that it couldn't persist but would continue.

File: activation/nipost.go
Comment (fasmat) [inline activation/nipost.go:230]: TODO: fetch from DB if available

File: activation/nipost.go
Comment (fasmat) [inline activation/nipost.go:277]: TODO: re-add missing data

File: activation/nipost.go
Comment (fasmat) [inline activation/nipost.go:435]: TODO: Add PoetRound ID to local state

File: activation/nipost.go
Comment (fasmat) [inline activation/nipost.go:444]: TODO: fetch PoetRound ID from local state

File: common/types/nipost.go
Comment (fasmat) [inline common/types/nipost.go:3]: TODO: consider moving them from here to `nipost_state` with the other deprecated functions

File: sql/migrations/local/0001_initial.sql
Comment (fasmat) [inline sql/migrations/local/0001_initial.sql:14]: TODO: use this field

File: sql/migrations/local/0001_initial.sql
Comment (fasmat) [inline sql/migrations/local/0001_initial.sql:17]: TODO: use this field

File: activation/nipost_test.go
Comment (fasmat) [inline activation/nipost_test.go:72]: TODO: complete test

File: activation/nipost.go
Comment (fasmat) [inline activation/nipost.go:58]: TODO: add migration code from old state to new

File: activation/nipost.go
Comment (poszu) [inline activation/nipost.go]: It seems that the code always reacts on error with a panic. How about inlining the err check + panic into `nipost.PoetRegistrationCount()`, perhaps naming it `MustPoet....()`?

File: common/types/nipost.go
Comment (poszu) [inline common/types/nipost.go]: I think you could remove `PoetServiceID` as well (along with the related getter in `poetClient interface`.

File: sql/localsql/0003_migration.go
Comment (poszu) [inline sql/localsql/0003_migration.go:118]: I'm not sure about this. All it takes is one poet client down or unresponsive and it will fail migration. What if we roll this out and then one of the poets crashes completely?

IMHO, to make this migration reliable we should first change the way poets are configured. Currently the `BaseConfig` takes just a slice of poet addresses:
```go
PoETServers []string `mapstructure:"poet-server"`
```
How about we change it to take an address and a poet's public key?
```go
type PoetServer struct {
	Address	string		`mapstructure:"address"`
	Pubkey	Base64Enc	`mapstructure:"pubkey"`
}

type BaseConfig struct {
...
	PoetServers []PoetServer `mapstructure:"poets"`
}
```
and we stop querying the poet service for its pubkey?

File: activation/nipost.go
Comment (fasmat) [reply activation/nipost.go]: An error here means that there is some issue that the program cannot easily recover from (SQL query cannot be parsed, DB connection dropped, etc.) and it makes sense to exit the program. I'm not a big fan of moving the panic into the function. `PoetRegistrationCount` is only called in two places and to safe this check would require passing the logger to the DB (which we do not do anywhere else).

File: common/types/nipost.go
Comment (fasmat) [reply common/types/nipost.go]: At the moment I still need the getter for the migration from state-files to DB. (see also my comment there)

File: sql/localsql/0003_migration.go
Comment (fasmat) [reply sql/localsql/0003_migration.go:118]: A failing migration should revert all changes that happened during the DB migration. On that note I probably need to not delete the migrated `*.bin` files but rather rename them (`*.bak`) and only delete them after the DB migration successfully completes (or leave them there in case we want a way to be able to downgrade if something goes wrong).

> What if we roll this out and then one of the poets crashes completely?

Ideally users can revert back to 1.2.6 and continue without issues.

EDIT: I also don't think the whole network will update at once. We don't have to force this update. Even if we would it will probably take at least an hour for most nodes to update and we are only querying `/info` which I thought was cached?

I agree we shouldn't fail if any request fails, but maybe only if ALL fail (the same way we do in NIPostBuilder). This however has the drawback that some PoETs might be dropped during migration that will then be online again during the cyclegap 😕.

> IMHO, to make this migration reliable we should first change the way poets are configured.

This will probably break things. Changes to the config are not trivial in the setup we are currently using, since the config is overwritten by Smapp from the config here: https://configs.spacemesh.network/config.mainnet.json So any change we make must work with both old and new nodes, i.e. we need to have poets twice in the config (old and new layout).

I think maybe we should just hard code a map of the official PoETs (Key -> URL) and provide a way for users to provide such a mapping for their custom PoETs if needed; but not via the config but a new file that is only read during migration and can be discarded afterwards.

File: activation/nipost.go
Comment (poszu) [reply activation/nipost.go]: Right, hence the panic - I'm totally fine with it. I'm only proposing to move the panic inside the function. You don't need to pass the logger to do that - just call `panic(...)`. IMHO it makes more sense to handle bugs like a broken SQL query inside of the function rather than leaking this to the caller (who has no idea why it could fail).

File: common/types/nipost.go
Comment (poszu) [reply common/types/nipost.go]: Right, but you moved the other ones into the migration files. Why keep this one here?

File: sql/localsql/0003_migration.go
Comment (poszu) [reply sql/localsql/0003_migration.go:118]: By "poet crashes completely" I meant a situation when one of the poets dies and doesn't boot anymore, and needs maintenance. In such a situation, all nodes would fail the migration until we fix the poet (if we even can fix it - it could take a week or so). In effect, the upgrade would be blocked until we fix the poet, which IMHO is a no-go.

---

>>   IMHO, to make this migration reliable we should first change the way poets are configured.
>
> This will probably break things. Changes to the config are not trivial in the setup we are currently using, since the config is overwritten by Smapp from the config here: https://configs.spacemesh.network/config.mainnet.json So any change we make must work with both old and new nodes, i.e. we need to have poets twice in the config (old and new layout).

Are you sure? Can't Smapp use config **A** for node version **a** and config **B** for node version **b**? Even if so, I think having poets temporarily configured twice (two separate fields) in the ` https://configs.spacemesh.network/config.mainnet.json` is not a big deal. It doesn't mean that the config structure of the new node version needs both. It will only contain the new field and the older one should be omitted during parsing.

We should communicate upgrade steps properly in the release nodes with an instruction on how to migrate the config in case somebody overwrites the list of poet servers (the default config would contain the list of SM poets with their pubkeys).

I can create a PR to change poet configuration as described.

File: common/types/nipost.go
Comment (fasmat) [reply common/types/nipost.go]: I thought that it would cause cyclic dependencies which doesn't seem to be the case so I moved it. 🙂 

With `v1.4.0` I would like to get rid of all the old code handling state files (`nipost_state*.go` in the `localsql` package) and also remove the data migration from disk to DB. New nodes don't need it and old nodes can upgrade `v1.2.x` -> `v1.3.x` -> `v1.4.x` to keep their state. We can then also remove the `PoetServiceID` method on `HTTPPoetClient`.

Wdyt?

File: activation/nipost.go
Comment (fasmat) [reply activation/nipost.go]: Interestingly, we seem to not have used `select count(*)` before in any other query, so there is no precedence to compare against.

I'm a bit hesitant to make `PoetRegistrationCount` directly panic on an error, because most if not all DB queries currently handle errors from `db.Exec` by just wrapping and returning them to the caller. The only way `db.Exec` however can fail is by either a) the connection to the DB was closed or b) the query is syntactically wrong. In both cases it makes no sense to continue and the node should just panic in my opinion; not just here.

I think the better way is to change `db.Exec` to not return an error and panic instead - to be consistent across all DB queries and if its OK for you I would do that in a dedicated PR.

File: sql/localsql/0003_migration.go
Comment (fasmat) [reply sql/localsql/0003_migration.go:118]: I see three options:

- fail the migration if a PoET isn't available while the migration happens. This will revert all changes and users can try again or users can downgrade to v1.2.6 and try upgrading later. As you said that's probably not the best solution since worst case scenario people might need to wait a full epoch to be able to upgrade (then the PoET wouldn't be in the new `nipost_builder_state.bin` file any more if it is still down)
- not fail the migration if at least one PoET is available when the migration happens. This would match the behaviour that the `NIPostBuilder` uses to fetch PoET proofs and would be my preferred solution. A PoET that is down (error or maintenance) is unlikely to have the highest tick count anyway, so losing that registration might not even matter.
- change the config to include both the URL and the public key of a PoET and force users to update their config accordingly. I would like to avoid that change, people might use the wrong public keys for the nodes (copy & paste errors, typos, swap keys accidentally) which can lead to multiple problems that might not surface until the cyclegap.

> Can't Smapp use config A for node version a and config B for node version b?

Afaik at the moment it cannot; also a lot of users don't use Smapp.

> We should communicate upgrade steps properly in the release nodes with an instruction on how to migrate the config in case somebody overwrites the list of poet servers (the default config would contain the list of SM poets with their pubkeys).

That could work if you think that the automatic "at least one PoET" migration is too risky.

Wdyt?

File: activation/nipost.go
Comment (fasmat) [reply activation/nipost.go]: I created the following issue to keep track of the change: https://github.com/spacemeshos/go-spacemesh/issues/5298

Comment (dshulyak) [reply activation/nipost.go]: it is not the exhaustive list. it can also exist with https://www.sqlite.org/rescode.html#busy depending on how it is used

Comment (fasmat) [reply activation/nipost.go]: Changed to instead return the error (so it will be logged and a bit later retried).

File: sql/localsql/0003_migration.go
Comment (poszu) [reply sql/localsql/0003_migration.go:118]: I finished and opened PR allowing to configure poets' public keys: https://github.com/spacemeshos/go-spacemesh/pull/5293.

Regarding:
>>Can't Smapp use config A for node version a and config B for node version b?
>
>Afaik at the moment it cannot; also a lot of users don't use Smapp.

@nj confirmed that it can serve different configs based on node version.

Comment (fasmat) [reply sql/localsql/0003_migration.go:118]: @poszu Updated the PR after #5293 was merged. It now uses the public keys from the config instead of querying them.

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

File: CHANGELOG.md
Comment (poszu) [inline CHANGELOG.md]: This probably needs to be relocated to either v1.3.1 or _unreleased_.

File: activation/e2e/nipost_test.go
Comment (poszu) [inline activation/e2e/nipost_test.go]: Why does the challenge need to be added?

File: activation/poet.go
Comment (poszu) [inline activation/poet.go]: How about not depending on `localsql` here and just use `[]byte`?

File: sql/localsql/0003_migration_interfaces.go
Comment (poszu) [inline sql/localsql/0003_migration_interfaces.go:9]: I think there's no need to export it.
```suggestion
type poetClient interface {
```

File: sql/localsql/nipost_state.go
Comment (poszu) [inline sql/localsql/nipost_state.go]: The `saveBuilderState` and `discardBuilderState` are only used in tests. How about moving them to _test.go?

**Review by poszu — APPROVED** [review]

File: sql/localsql/0003_migration_interfaces.go
Comment (fasmat) [reply sql/localsql/0003_migration_interfaces.go:9]: I tried to have it private, but I need it in `node.go` (around line 1620) when I instantiate the poet clients and pass them to the migration that needs it. Since the `node` package already imports `localsql` this made sense to me but wdyt?

File: sql/localsql/nipost_state.go
Comment (fasmat) [reply sql/localsql/nipost_state.go]: `saveBuilderState` is only used by the test and I can move it there, but `discard-` and `loadBuilderState` are needed by the migration.

File: activation/e2e/nipost_test.go
Comment (fasmat) [reply activation/e2e/nipost_test.go]: seems to be a leftover

**Comment by fasmat** [comment]
bors try

File: sql/localsql/0003_migration_interfaces.go
Comment (poszu) [reply sql/localsql/0003_migration_interfaces.go:9]: I don't understand why the interface needs to be exported. The poet HTTP client from the `activation` pkg satisfies it regardless (of exported/private) so it should work. Why not?

Comment (fasmat) [reply sql/localsql/0003_migration_interfaces.go:9]: I tried it, but the call in `node.go` takes a slice of interfaces and Go doesn't support covariance on slices. I would need to make `New0003Migration` take `[]*activation.HTTPClient` if we don't export the interface, but that would complicate testing and requires some refactoring (to avoid cyclic dependencies).

**Comment by fasmat** [comment]
bors merge

**Pull request closed without merging**
